systemFlake.nix

```diff
@@ -86,24 +86,51 @@ let
     let
       selectedNixpkgs = getNixpkgs host;
       host = evalHostArgs (mergeAny hostDefaults host');
+      channelValue = channels.${host.channelName};
+      patchedChannel = patchChannel host.system channelValue.input (channelValue.patches or [ ]);
+      # Use lib from patched nixpkgs
+      lib = import (patchedChannel + "/lib");
+      # Use nixos modules from patched nixpkgs
+      baseModules = import (patchedChannel + "/nixos/modules/module-list.nix");
+      # Override `modulesPath` because otherwise imports from there will not use patched nixpkgs
+      specialArgs = { modulesPath = builtins.toString (patchedChannel + "/nixos/modules"); } // host.specialArgs;
+      # The only way to find out if a host has `nixpkgs.config` set to
+      # the non-default value is by evalling most of the config.
+      hostConfig = (lib.evalModules {
+        prefix = [ ];
+        check = false;
+        modules = baseModules ++ host.modules;
+        args = { inherit inputs; } // host.extraArgs;
+        inherit specialArgs;
+      }).config;
     in
     {
-      ${host.output}.${hostname} = host.builder {
-        inherit (selectedNixpkgs) system;
+      ${host.output}.${hostname} = host.builder ({
+        inherit (host) system;
         modules = [
-          ({ pkgs, lib, options, ... }: {
+          ({ pkgs, lib, options, config, ... }: {
             # 'mkMerge` to separate out each part into its own module
             _type = "merge";
             contents = [
               (optionalAttrs (options ? networking.hostName) {
                 networking.hostName = hostname;
               })
 
-              (if options ? nixpkgs then {
-                nixpkgs = {
-                  inherit (selectedNixpkgs) overlays config system;
-                };
-              } else { _module.args.pkgs = selectedNixpkgs; })
+              (if options ? nixpkgs then
+              # Make sure we don't import nixpkgs again if not
+              # necessary. We can't use `config.nixpkgs.config`
+              # because that triggers infinite recursion.
+                if (hostConfig.nixpkgs.config == { }) then
+                  { nixpkgs.pkgs = selectedNixpkgs; }
+                else
+                  {
+                    nixpkgs.pkgs = import patchedChannel {
+                      inherit (host) system;
+                      overlays = sharedOverlays ++ (if (channelValue ? overlaysBuilder) then (channelValue.overlaysBuilder self.pkgs) else [ ]);
+                      config = channelsConfig // (channelValue.config or { }) // config.nixpkgs.config;
+                    };
```

Comment:
Consider this instead:
```nix
                {
                  nixpkgs.pkgs =
                    if (hostConfig.nixpkgs.config == { })
                    then selectedNixpkgs
                    else
                      import patchedChannel {
                        inherit (host) system;
                        overlays = sharedOverlays ++ (if (channelValue ? overlaysBuilder) then (channelValue.overlaysBuilder self.pkgs) else [ ]);
                        config = channelsConfig // (channelValue.config or { }) // config.nixpkgs.config;
                      };
                }
```

Reply:
Wait a second, won't this use a patched version of nixpkgs in case no configuration defined at all? :thinking:

Reply:
Both use patched nixpkgs, the difference is the config that is passed to the patched nixpkgs. In the first case the config is only the config set in flake.nix, in the second the configuration from the host's modules is taken into account as well.